Added a tutorial for simulating Ising 1-D chain using mitiq with cirq… (#2002)
unitaryfund/mitiq: docs/source/examples/quantum_simulation_1d_ising.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.15.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

> # Use CDR, ZNE, and VNCDR to mitigate the errors in the simulation of 1-D Transverse-Longitudinal Ising model

Reviewer: ```suggestion
# CDR, ZNE, and VNCDR: Simulating 1-D Transverse-Longitudinal Ising model
```

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.15.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

 # CDR, ZNE, and VNCDR: 1-D Transverse-Longitudinal Ising model

> In this tutorial, we employ ZNE, CDR, and VNCDR mitigation techniques to address errors in the simulation of the 1-D Transverse-Longitudinal Ising model using Mitiq. It is important to note that the results presented here are not original, but rather an attempt to reproduce some of the findings outlined in the paper available at {cite}`Sopena_2023_Quantum`

Reviewer 2: ```suggestion
In this tutorial, we employ ZNE, CDR, and VNCDR mitigation techniques to address errors in the simulation of the 1-D Transverse-Longitudinal Ising model using Mitiq. It is important to note that the results presented here are not original, but rather an attempt to reproduce some of the findings outlined in the paper available at {cite}`Sopena_2023_Quantum`.

```

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.15.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

>  # CDR, ZNE, and VNCDR: 1-D Transverse-Longitudinal Ising model

Reviewer 2: ```suggestion
 # Using ZNE and learning-based methods to mitigate the 1D transverse-longitudinal Ising model
```

# Invert the y-axis
plt.gca().invert_yaxis()

# Add a title
plt.title('Mitigated CDR')

# Add x and y labels
plt.xlabel(r'$\Delta_i$')
plt.ylabel('n_dt')

# Add a colorbar for reference
plt.colorbar()

# Show the plot
plt.show()

# Plot mitigated_measurement_vncdr using colors
plt.imshow(np.transpose(mitigated_measurement_vncdr), cmap='viridis')

# Invert the y-axis
plt.gca().invert_yaxis()

# Add a title
plt.title('Mitigated VNCDR')

# Add x and y labels
plt.xlabel(r'$\Delta_i$')
plt.ylabel('n_dt')

# Add a colorbar for reference
plt.colorbar()

# Show the plot
plt.show()
```

The plots are consist of different cells, and the color of each cell represents the value of $\Delta_i$ at that particular time.

To enhance the visualization of the performance of various mitigation techniques, the absolute value of the difference between the exact measurement and different mitigation techniques is presented.

```{code-cell}

# Plot the difference between exact and unmitigated_measurement using colors
plt.imshow(np.abs(np.transpose(exact_measurement)-np.transpose(unmitigated_measurement)), cmap='viridis')

# Invert the y-axis
plt.gca().invert_yaxis()

# Add a title

> plt.title('Differnce between exact Trotterization and unmitigated')

Reviewer 2: ```suggestion
plt.title('Difference between exact Trotterization and unmitigated')
```

# Add x and y labels
plt.xlabel(r'$\Delta_i$')
plt.ylabel('n_dt')

# Add a colorbar for reference
plt.colorbar()

# Show the plot
plt.show()

# Plot the difference between exact and mitigated_measurement_zne using colors
plt.imshow(np.abs(np.transpose(exact_measurement)-np.transpose(mitigated_measurement_zne)), cmap='viridis')

# Invert the y-axis
plt.gca().invert_yaxis()

# Add a title

> plt.title('Differnce between exact Trotterization and mitigated ZNE')

Reviewer 2: ```suggestion
plt.title('Difference between exact Trotterization and mitigated ZNE')
```

# Add x and y labels
plt.xlabel(r'$\Delta_i$')
plt.ylabel('n_dt')

# Add a colorbar for reference
plt.colorbar()

# Show the plot
plt.show()

# Plot the difference between exact and mitigated_measurement_cdr using colors
plt.imshow(np.abs(np.transpose(exact_measurement)-np.transpose(mitigated_measurement_cdr)), cmap='viridis')

# Invert the y-axis
plt.gca().invert_yaxis()

# Add a title

> plt.title('Differnce between exact Trotterization and mitigated CDR')

Reviewer 2: ```suggestion
plt.title('Difference between exact Trotterization and mitigated CDR')
```

# Add x and y labels
plt.xlabel(r'$\Delta_i$')
plt.ylabel('n_dt')

# Add a colorbar for reference
plt.colorbar()

# Show the plot
plt.show()

# Plot the difference between exact and mitigated_measurement_vncdr using colors
plt.imshow(np.abs(np.transpose(exact_measurement)-np.transpose(mitigated_measurement_vncdr)), cmap='viridis')

# Invert the y-axis
plt.gca().invert_yaxis()

# Add a title

> plt.title('Differnce between exact Trotterization and mitigated VNCDR')

Reviewer 2: ```suggestion
plt.title('Difference between exact Trotterization and mitigated VNCDR')
```

mitigated_measurement_zne=[1]

# repeat the trotter evolution n_dt times and compute compute the unmitigated, exact and mitigated quantities for each step using Mitiq
for ii in range(n_dt):
  test_circuit+=trotter_evolution_H(L,h_z,h_x,Jdt)
  unmitigated_measurement.append(obs.expectation(test_circuit, noisy_simulator).real)
  exact_measurement.append(obs.expectation(test_circuit,exact_simulator))
  mitigated_measurement_vncdr.append(cdr.execute_with_cdr(
    test_circuit,
    noisy_simulator,
    observable=obs,
    simulator=exact_simulator,
    scale_factors=(1,3)
).real)
  mitigated_measurement_cdr.append(cdr.execute_with_cdr(
    test_circuit,
    noisy_simulator,
    observable=obs,
    simulator=exact_simulator,
).real)
  mitigated_measurement_zne.append(zne.execute_with_zne(
    test_circuit,
    noisy_simulator,
    observable=obs,
).real)

  # this is to keep track of how fast the simulation is
  print(ii)

```

+++ {"id": "xXoS2SuzDCqX"}

We plot all the exact, unmitigated, and different mitigated measurements to compare the results. As suggested in the original paper https://arxiv.org/pdf/2103.12680.pdf , the evolution of the $<Z_2(t)>$ is as follows

$

> <σ^z_i(t)>=a_1 e^{-a_2 t}cos(a_3t)+a_4t+a_5

Reviewer 2: please check also elsewhere where these changes need to be made.

$

we used the trotterized simulation results to estimate parametrs $a_1$,...,$a_5$.

```{code-cell}
---
colab:
  base_uri: https://localhost:8080/
  height: 450
id: Y9xQMpwMOhI1
outputId: 62ac6547-c275-4007-d62e-e410de58a817
---
import numpy
from scipy.optimize import curve_fit
import matplotlib.pyplot as plt

# Define the average value function
def sigma_z_t(x, a_1, a_2, a_3, a_4, a_5):
    return a_1 * numpy.exp(-a_2 * x) * numpy.cos(a_3 * x) + a_4 * x + a_5

# Generate some sample data
x_data = np.linspace(0, n_dt*Jdt, n_dt+1)

# Fit the cosine function to the data
popt, pcov = curve_fit(sigma_z_t, x_data, exact_measurement)

# Extract the optimized parameters
a_1_opt, a_2_opt, a_3_opt, a_4_opt, a_5_opt = popt

# Generate the fitted curve
x_fit = np.linspace(0, n_dt*Jdt, 300)

> y_fit = sigma_z_t(x_fit, *popt)

Author: do you mean to do exact trotterized results? since I do not know how to simulate that Hamiltonian other than trattorization. I used two different names "exact" and "trotterized", I should use the same name for them, something like "exact trotterization".

let me know if this is what you are suggesting.
